pkg/tls/cred.go

```diff
@@ -90,6 +91,15 @@ func (crt *Crt) Verify(roots *x509.CertPool, name string) error {
 	}
 	vo := x509.VerifyOptions{Roots: roots, Intermediates: i, DNSName: name}
 	_, err := crt.Certificate.Verify(vo)
+
+	if crtExpiryError(err) {
+
+		//need to remove seconds and nanoseconds for testing returned error
```

Comment:
so one way to make this testable without needing to throw away precision in the actual codepath would be to define a new error struct that this method could return.  The struct would implement the error interface and would hold a Time field.  The Error() method would do the formatting that you have here.

The advantage of this is that when we call `Verify` in our tests, we can then cast the error to our known error struct, extract the Time object, and compare it to the current Time with a certain error tolerance.

Reply:
That's a great idea @adleong ! I've made this change, could you take a look?

Reply:
@adleong @mayankshah1607  Just took a look and I think there is a better way to do this. So currently the verify method looks like: 

```go
func (crt *Crt) Verify(roots *x509.CertPool, name string) error {
	i := x509.NewCertPool()
	for _, c := range crt.TrustChain {
		i.AddCert(c)
	}
	vo := x509.VerifyOptions{Roots: roots, Intermediates: i, DNSName: name}
	_, err := crt.Certificate.Verify(vo)

	if _, ok := crtExpiryError(err, *crt.Certificate); ok {
		return fmt.Errorf("%s - Current Time : %s - Invalid before %s - Invalid After %s", err, time.Now(), crt.Certificate.NotBefore, crt.Certificate.NotAfter)
	}
	return err
}
```

we can change the signature to be `func (crt *Crt) Verify(roots *x509.CertPool, name string, currentTime *time.Time)`. This will allow us to actually insert this time (if not nil into ) VerifyOptions.CurrentTime. Much more predictable than trying to shave off precision and fight the wall clock. WDYT ?

Reply:
@mayankshah1607 Here is a [diff](https://gist.github.com/zaharidichev/0ec8d4a02fac5becf555e7718893d4a8) pointing to what I mean if you want to explore that avenue

Reply:
@zaharidichev this idea too looks good. I did try something like this before, but had to discard it because of the refactoring that would follow. But if that's not an issue, I think we could very well go ahead with this approach. @adleong WDYT? I'll push these changes right away if I can have your :+1:

Reply:
Although one small concern - I think the `x509.Verify` method calls `time.Now` independent of it's parent `Verify` method, which may lead to inconsistency of nanoseconds (which probably might not matter); i.e `time.Now` which actually triggers the expiry error is a few nanoseconds ahead of what's displayed on the log message.

Reply:
@mayankshah1607 if you are refering to this here: https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L576, the method is called only if the CurrentTime in opt is a zero value one, meaning that if you supply our own CurrentTime in the options, it would use that. Am I missing something ?

Reply:
Got it! Did not strike me that we could just add the time into `opts` from the func args. :+1:

Reply:
This is a good idea!